packages/grafana-ui/src/components/GraphNG/utils.ts

```diff
@@ -1,31 +1,152 @@
-  if (!data.length) {
+export function mergeDataFrames(frames: DataFrame[]): AlignedFrameWithGapTest | null {
+  let valuesFromFrames: AlignedData[] = [];
+
+  for (let i = 0; i < frames.length; i++) {
+    const frame = frames[i];
+
+    let { timeField } = getTimeField(frame);
+
+    // walk all temporal frames
+    if (timeField !== undefined && frame.fields.length > 1) {
+      // push time values
+      let alignedData: AlignedData = [timeField.values.toArray()];
+
+      // push numeric values
+      for (let j = 0; j < frame.fields.length; j++) {
+        const field = frame.fields[j];
+
+        if (field.type === FieldType.number) {
+          let values = field.values.toArray();
+
+          if (field.config.custom?.nullValues === 'asZero') {
+            values = values.map(v => (v === null ? 0 : v));
+          }
+
+          alignedData.push(values);
+        }
+      }
+
+      valuesFromFrames.push(alignedData);
+    }
+  }
+
+  if (valuesFromFrames.length === 0) {
     return null;
   }
 
-  // normalize time field names
-  // in each frame find first time field and rename it to unified name
-  for (let i = 0; i < data.length; i++) {
-    const series = data[i];
-    for (let j = 0; j < series.fields.length; j++) {
-      const field = series.fields[j];
-      if (field.type === FieldType.time) {
-        field.name = fieldName;
-        break;
+  // do the actual alignment (outerJoin on the first arrays)
+  let { data: alignedData, isGap } = outerJoinValues(valuesFromFrames);
+
+  // the outerJoined frame we're gonna poop out
+  let alignedFrame: DataFrame = {
+    name: '<multiple>',
+    refId: '<multiple>',
+    fields: [],
+    length: alignedData[0].length,
+  };
+
+  // populate the alignedFrame with original Fields but with aligned values
+  for (let i = 0, seriesIdx = 0; i < frames.length; i++) {
+    const frame = frames[i];
+
+    let { timeField } = getTimeField(frame);
+
+    // walk all temporal frames
+    if (timeField !== undefined && frame.fields.length > 1) {
+      // push time field
+      if (alignedFrame.fields.length === 0) {
+        alignedFrame.fields.push({ ...timeField, values: new ArrayVector(alignedData[seriesIdx++]) });
+      }
+
+      // push numeric fields
+      for (let j = 0; j < frame.fields.length; j++) {
+        const field = frame.fields[j];
+
+        if (field.type === FieldType.number) {
+          alignedFrame.fields.push({ ...field, values: new ArrayVector(alignedData[seriesIdx++]) });
+        }
+      }
+
+      valuesFromFrames.push(alignedData);
+    }
+  }
+
+  return {
+    frame: alignedFrame,
+    isGap,
+  };
+}
+
+export function outerJoinValues(tables: AlignedData[]): AlignedDataWithGapTest {
+  if (tables.length === 1) {
+    return {
+      data: tables[0],
+      isGap: () => true,
+    };
+  }
+
+  let xVals: Set<number> = new Set();
+  let xNulls: Array<Set<number>> = [new Set()];
+
+  tables.forEach(t => {
```

Comment:
not a strict rule but we try to use `for (const table of tables) {`  instead of forEach with a function callbacks, mainly because they are faster, supports break & return & continue like normal loops

Reply:
sure thing. i'm still in the habit of avoiding constructs which have heavy down-transpiles to es5: https://babeljs.io/docs/en/babel-plugin-transform-for-of

i think i'll drop all es5-isms in uPlot 1.5 [1]. i'm quite sure uPlot doesnt work with IE11 any more anyhow, nor should it.

[1] https://github.com/leeoniya/uPlot/blob/0316e6b5755ba1bd5d9607d02923b55a163d96e2/rollup.config.js#L16